libraries/GCS_MAVLink/MAVLink_routing.cpp

```diff
@@ -94,8 +94,31 @@ detect a reset of the flight controller, which implies a reset of its
 routing table.
 
 */
-bool MAVLink_routing::check_and_forward(GCS_MAVLINK &in_link, const mavlink_message_t &msg)
+bool MAVLink_routing::check_and_forward(uint8_t framing_status,
+                                        GCS_MAVLINK &in_link,
+                                        const mavlink_status_t &status,
+                                        const mavlink_message_t &msg)
 {
+    // handle the case of loopback of our own messages, due to
+    // incorrect serial configuration.
+    if (msg.sysid == mavlink_system.sysid &&
+        msg.compid == mavlink_system.compid) {
+        return false;  // do not process locally
+    }
+
+    switch (framing_status) {
+    case MAVLINK_FRAMING_OK:
+        break;
+    case MAVLINK_FRAMING_BAD_CRC:
+        if (in_link.option_enabled(GCS_MAVLINK::Option::FORWARD_BAD_CRC)) {
+            forward(in_link, msg);
+        }
+        return false;  // do not process locally
+    case MAVLINK_FRAMING_BAD_SIGNATURE:
```

Comment:
need to test that the USB override of bad signature to allow fixing of signing key on USB still works